odoo/addons/base/models/res_users.py

```diff
@@ -394,9 +394,50 @@ def _check_action_id(self):
     @api.multi
     @api.constrains('groups_id')
     def _check_one_user_type(self):
-        for user in self:
-            if len(user.groups_id.filtered(lambda x: x.category_id.xml_id == 'base.module_category_user_type')) > 1:
-                raise ValidationError(_('The user cannot have more than one user types.'))
+        """We check that no users are both portal and users (same with public).
+           This could typically happen because of implied groups.
+        """
+        user_types_category = self.env.ref('base.module_category_user_type', raise_if_not_found=False)
+        user_types_groups = self.env['res.groups'].search(
+            [('category_id', '=', user_types_category.id)]) if user_types_category else False
+        if user_types_groups:  # needed at install
+            bad_users = self._fast_check_users_in_more_than_one_group(user_types_groups.ids)
+            if bad_users:
+                raise ValidationError(_('The user cannot have more than one user types.') + '\n' +
+                                      _('Problematic ids: ') + ", ".join(str(bad[0]) for bad in bad_users))
+
+    @api.multi
+    def _fast_check_users_in_more_than_one_group(self, group_ids):
+        """The method is not 'fast' if the list of ids is very long;
+           e.g. using group.users as self with 10^5 users makes this query take 1 minute
+           that's why we need to split the cases depending on the user ids size;
+           when called from groups we use a subquery to get the ids from the groups,
+           and as backup we check everything
+        :param group_ids: list of group ids
+        :optional context param from_groups: ids of groups on which we want to check the users
+        :return: the users that are in more than one of the provided groups
+        """
+        if group_ids and (self or self.env.context.get('from_groups')):
+            args = [tuple(group_ids)]
+            if self.env.context.get('from_groups'):  # should be the ids of the groups we are interested in
```

Comment:
If the check on all users is *fast* maybe we can get rid of the `from_groups` and the `len(self.ids) < 10000`. As you mentioned, that means that we check all groups and not only the one we change.

Reply:
I've been doing some experiments on a test DB with 1M users and 5M group memberships.

First of all, the main query could be surrounded with a `SELECT EXISTS(...)`. This will make it much faster when there *is* a dual membership, as it won't need to compute the whole aggregation table, it will stop at the first one. When one group involved has 1M users it can reduce the exec time from *several seconds* to 0.5ms!
It means we cannot report the conflicting users in the error, but do we really need that?

Then, several groups with millions of users are extremely unlikely, especially for the specific groups we're checking. They're normally mutually exclusive and one of the groups is usually much larger than the others (e.g. portal).
So the worst case is when there is no dual membership, because the whole table needs to be aggregated to return the result. In that case an unfiltered query (with no filter on users) can take several hundred milliseconds (600-700ms in my sample db). This is a bit prohibitive when we're in fact testing only a single user: if we filter the same query on a single user it executes in 0.5 ms!

So there *is* a good reason to filter on the users.

But maybe we don't need to have complicated thresholds and `from_groups`. In reality there are 2 main use cases:
 - editing a single user in the UI - the query should be very fast in all cases
 - editing groups of users, or mass editing users via RPC - it won't surprise anyone if it's a bit slow on large databases (multi millions users)

I don't think we should care much about intermediary cases like "I'm changing the groups of just 2 users". It's infrequent and can be seen as a "mass editing" case.

So I suggest we handle *only* those 2 cases? If `len(self.id) == 1` we add a filter with `r.uid = %s`. And in all other cases we do the completely unfiltered query.

Looks like a good compromise:
- single-user check is very fast (<1ms)
- multi-user check is relatively fast (<1s on my 1M users DB), and executes in constant time regardless of number of users tested
- no need for weird thresholds
- `from_group` becomes un-necessary because we never pass huge lists of IDs anymore
- bonus: the code will be trivial

What do you think?

Reply:
Before writing the complicated version, I hesitated on settling on something close to that.
It's a trade-off between functionality and code complexity, because in the last version you don't even know if it's the last action you performed that caused the error or if it was already there (e.g. in case of migrated/old databases). 
So in some cases there will be a bit more work in understanding what went wrong, but I guess it's acceptable. People who won't be able to do it themselves will complain to support who should not have trouble with it.
 
So I'm fine with it, and implemented all the proposed changes.